File: frontend/src/modules/onespace/lib/shell/assistant.js
/**
 * Whether the workspace has an assistant, and which conversations it holds.
 *
 * Two readers and neither owns the other: the rail decides whether to draw the
 * entry at all, and the page draws the list of threads. A module-level
 * reactive for the same reason `mail.js` is one — the rail outlives the page,
 * so there is no parent to pass it down from.
 *
 * Fetched once rather than polled. Whether AI is switched on for a workspace
 * changes when somebody changes it, which is a page reload away; a minute
 * timer would be a request a minute to answer a question whose answer does not
 * move.
 */
import { computed, reactive } from 'vue'
import { close, onDesk, open, press } from '@/modules/onespace/lib/desk/windows'
import { assistant as booted } from '@/shared/lib/runtime/boot'
import { workspace } from '@/shared/lib/workspace'
import { __ } from '@/shared/lib/runtime/translate'

/** Which window on the desk this is. One id, said once. */
export const ASSISTANT = 'assistant'

export const assistant = reactive({
  // What it is looking at. Here rather than on the panel itself because the
  // thing that opens it is somewhere else every time — the dock, a record's
  // controls, a keyboard shortcut — and none of them is its parent.
  //
  // Whether it is *showing* is no longer here: the assistant is one window
  // among several now, and the desk keeps the list. Two places saying whether
  // it is open is how one of them comes to be wrong — the dock would light a
  // tile for a panel nobody had opened.
  on: null,
  session: '',
  // False until we know. The rail draws nothing while it is false, which is
  // the right way round: an entry that appears and then disappears is worse
  // than one that appears a beat late.
  available: false,
  sessions: [],
  loaded: false,
  // Who it is, from the boot payload. Here rather than fetched with the thread
  // list, because the rail names it before that call returns and a label that
  // says "Assistant" for a beat and then says something else is the same
  // visible flicker the favicon had. `www/one.py` puts it there.
  name: booted?.name || '',
  avatar: booted?.avatar || '',
})

/**
 * Whether its window is on the desk. Read by the dock, which lights the tile,
 * and by the widget, which draws the panel.
 *
 * On the desk rather than *drawn*: a folded assistant is still open, and its
 * tile says so. `DeskWindow` is what decides whether a folded window is
 * painted.
 */
export const assistantShowing = computed(() => onDesk(ASSISTANT))

/** What to call it. Never empty — a nameless assistant is still on screen, and
 *  this is the same fallback the server uses when nothing has been set. */
export const assistantName = computed(() => assistant.name || __('Assistant'))

/** Its picture, or nothing: `Avatar` draws a letter from the label instead. */
export const assistantAvatar = computed(() => assistant.avatar)

/**
 * What the AI settings tab calls once a save has come back.
 *
 * Without it the rail and the panel keep the old name until a reload, which is
 * the kind of wrong nobody reports and everybody notices.
 */
export function setAssistant(who) {
  assistant.name = who?.name || ''
  assistant.avatar = who?.avatar || ''
}

/**
 * Open the panel, against what the caller is looking at.
 *
 * `on` is `{space, screen, docname, label}` or nothing. Only the label is for
 * the browser; the rest goes to the server, which resolves every part of it
 * through the same checks a click goes through — so a context this reader
 * cannot reach narrows to nothing rather than widening anything.
 *
 * Changing what it is looking at starts a new thread. A conversation that was
 * about a quotation and is now about a project is two conversations, and
 * carrying the first one forward would leave the model answering "this one"
 * from a note that no longer applies.
 */
export function openAssistant(on = null) {
  if (JSON.stringify(on || null) !== JSON.stringify(assistant.on || null)) {
    assistant.on = on || null
    assistant.session = ''
  }
  open(ASSISTANT)
  loadAssistant()
}

export function closeAssistant() {
  close(ASSISTANT)
}

/**
 * What the dock's tile does, which is not quite opening.
 *
 * A tile is a toggle: shut it opens, in front it folds away, behind it comes
 * forward. Only the first of those is `openAssistant`, and only the first of
 * them should carry a subject — folding a window is not a statement about what
 * you want to ask, so passing the context through here would start a new
 * thread every time somebody put the panel away and got it back.
 */
export function pressAssistant(on = null) {
  if (onDesk(ASSISTANT)) press(ASSISTANT)
  else openAssistant(on)
}

/** Ask the server once whether the assistant is on, and for the thread list. */
export async function loadAssistant({ reload = false } = {}) {
  if (assistant.loaded && !reload) return assistant
  try {
    const got = await workspace.assistantSessions()
    assistant.available = !!got?.available
    assistant.sessions = got?.sessions || []
  } catch {
    // A workspace with no AI configured refuses rather than answering an empty
    // list, and the rail's answer to a refusal is the same as its answer to
    // "switched off": offer nothing.
    assistant.available = false
    assistant.sessions = []
  }
  assistant.loaded = true
  return assistant
}

File: oneapp/onespace/chat/assistant.py
"""The workspace assistant: one declared AI feature that happens to loop.

Everything about it is ordinary. It is an `@ai_feature`, so it appears in the
workspace's AI settings as a row with a model picker filtered to the capability
it declared, it can be switched off, a workspace may append to its instructions
and may never read them, and every turn holds credits before the call and
settles them after. There is no second AI path and no configuration surface of
its own — a chat that could name a provider would be the thing we do not do.

What makes it a chat rather than a summary is two lines of the declaration:
`tools` names what it may read, and `max_turns` says how many calls one question
may become. `ai/conversation.py` is the loop; `toolbox.py` is what the loop may
run; neither is reachable except through here.

The cost shape is the thing to keep in view. A question that needs three
lookups is four calls, four holds and four settlements, and it is charged as
four — which is honest and is also why `max_run_credits` exists.
"""

import frappe
from frappe import _

from oneapp.onespace.ai import conversation
from oneapp.onespace.ai.features import ai_feature
from oneapp.onespace.chat import context, session as store
from oneapp.onespace.chat.toolbox import tools

SYSTEM = """You are the assistant inside a OneSpace workspace. You help the \
person you are talking to with their own work in this workspace.

Answer from what the tools tell you. You cannot see anything until you look, so \
when a question is about this workspace's records, files or documents, use the \
tools rather than answering from memory — and when the tools return nothing, \
say that nothing was found rather than filling the gap.

The tools run as the person asking, so what you can read is what they could \
have opened themselves. If a tool refuses, tell them plainly that they do not \
have access to it; never speculate about what the record might have said.

Work in the workspace's own words. A screen is called what its label calls it, \
and a record is one of whatever the screen says it is one of. The names the \
tools take and return are for the tools; the answer uses the words on screen.

Be brief. Give the answer first. Quote figures exactly as the tools reported \
them and never round a total or estimate a count — call count_records instead. \
When you name a record, give the id the tools returned so it can be found.

You cannot change anything. Every tool is read-only. If you are asked to create, \
edit, send or delete something, say that you can only read, and say where in the \
workspace it can be done."""


@ai_feature(
	"chat.workspace",
	label="Workspace assistant",
	capability="Text Generation",
	system=SYSTEM,
	description="Answers questions about this workspace's records, files and documents.",
	tools="oneapp.onespace.chat.toolbox.tools",
	# One question, at most eight calls. Enough for look, narrow, read, answer
	# with room to recover from a wrong guess; short enough that a model stuck
	# in a loop costs a known amount rather than an open-ended one.
	max_turns=8,
	max_run_credits=25,
	# Per call, and generous on input because a transcript grows: by the sixth
	# turn the prompt carries five answers and everything they returned.
	max_input_tokens=120_000,
	max_output_tokens=2_000,
)
def ask(ai, session: str, question: str, on: dict | None = None) -> dict:
	"""Answer one question in a stored conversation, using the workspace's data.

	`on` is what the reader has open, already checked by `context.read`. It
	narrows the tools to that space and puts one sentence in front of the model
	saying which screen and which record — so "is this priced above the last
	one?" works in the panel beside a quotation and means nothing from the rail.
	"""
	on = on or {}
	spoken = store.transcript(session) + [{"role": "user", "content": question}]

	run = conversation.run(ai, spoken, context.bound(tools(), on), context.note(on))

	# Only what this ask added. `run.messages` is the whole transcript because
	# the loop needs it; storing it whole would write every earlier turn again.
	store.append(session, run["messages"][len(spoken) - 1:],
	             credits=run.get("credits") or 0, stopped=run.get("stopped") or "")
	return run


# --------------------------------------------------------------------------- #
# Endpoints
#
# Four, and none of them takes a model, a prompt or a limit. What the assistant
# is allowed to do is declared above; a browser picks the session and types the
# question.
# --------------------------------------------------------------------------- #

@frappe.whitelist(methods=["GET"])
def sessions() -> dict:
	"""The threads this person has, and whether the assistant is available."""
	from oneapp.onespace.ai import features, gateway

	feature = features.get("oneapp.chat.workspace")
	return {
		"sessions": store.listing(),
		"available": bool(feature and gateway.is_configured()
		                  and features.is_enabled(feature.key)),
	}


@frappe.whitelist(methods=["GET"])
def messages(session: str) -> dict:
	"""One thread, as it is shown: the turns a person sees and what was looked at.

	`tool` rows are not returned as messages. They are the working, not the
	conversation — but which tools ran is worth showing, so the lookups are
	carried onto the answer they led to and the reader can see it came from four
	quotations rather than from nowhere.

	A turn where the model only asked for tools has no text of its own. Those
	rows are not sent either: rendering one is an empty bubble between the
	question and the answer, which reads as a failure. Their lookups gather onto
	the next turn that actually says something, so a reply arrives with its whole
	working under it.
	"""
	shown, gathered = [], []
	for row in store.rows(session):
		if row["role"] == "tool":
			continue

		gathered += [
			{"tool": call.get("name") or "", "arguments": call.get("arguments") or {}}
			for call in store._parsed(row.get("tool_calls"))
		]

		# The exception is a run that stopped without answering: it has no text
		# either, and it is the one case where the silence has to be shown,
		# because `stopped` is what the reader is told instead.
		if not (row.get("content") or "").strip() and not row.get("stopped"):
			continue

		shown.append({
			"name": row["name"],
			"role": row["role"],
			"content": row.get("content") or "",
			"looked_at": gathered,
			"credits": row.get("credits") or 0,
			"stopped": row.get("stopped") or "",
			"on": str(row.get("creation") or ""),
		})
		gathered = []

	return {"session": session, "messages": shown}


@frappe.whitelist(methods=["POST"])
def send(question: str, session: str = "", on: str | dict | None = None) -> dict:
	"""Ask. Opens a thread if there is not one yet, and returns the whole reply.

	Not streamed. A streaming answer would have to hold a worker open for the
	length of a loop that may make eight provider calls, and a chat is not worth
	a request that can occupy a worker for two minutes. What arrives instead is
	the finished answer with the tools it used beside it.

	`on` is where the question was asked from — `{space, screen, docname}`, as
	the panel knows it. It arrives from a browser, so `context.read` resolves it
	through the same checks a click goes through and drops what does not hold.
	"""
	from oneapp.onespace.ai import features, gateway

	question = (question or "").strip()
	if not question:
		frappe.throw(_("Ask something first."))

	session = session or store.start(question)
	store.mine(session, "write")

	try:
		run = ask(session, question, context.read(on))
	except gateway.OutOfCredits as e:
		return {"session": session, "ok": False, "reason": "insufficient_credits",
		        "message": str(e)}
	except features.FeatureDisabled as e:
		return {"session": session, "ok": False, "reason": "disabled", "message": str(e)}

	return {
		"session": session,
		"ok": True,
		"reply": run.get("reply") or "",
		"stopped": run.get("stopped") or "",
		"credits": run.get("credits") or 0,
		"looked_at": run.get("tool_calls") or [],
		**messages(session),
	}


@frappe.whitelist(methods=["POST"])
def forget(session: str) -> dict:
	"""Delete a thread. There is no archive-and-keep: gone is what it says."""
	store.remove(session)
	return {"ok": True}
